### PR TITLE
Guide for AWS

### PR DESCRIPTION
**Pull request by gibbster**
Updated this to use main branch. Replaces https://github.com/swift-server/guides/pull/17

### CODE REVIEW

File: docs/deployment.md [hunk 1]
```diff
@@ -4,8 +4,9 @@
 The following guides can help with the deployment to public cloud providers:
 * [DigitalOcean](digital-ocean.md)
 * [Heroku](heroku.md)
+* [AWS](aws.md)
```
Comment (tomerd) [inline docs/deployment.md:7]: shell we sort these alphabetically?

File: docs/aws.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+Finally, check that Swift is correctly installed by running the Swift REPL: `swift`.
+
+![Invoke REPL](../images/aws/repl.png)
+
+Let's now download and build an test application. We will use the `--static-swift-stdlib` option so that it can be deployed to a different server without the Swift toolchain installed.
+
+```
+git clone https://github.com/apple/swift-nio.git
+cd swift-nio
+swift build -v --static-swift-stdlib -c release
```
Comment (tomerd) [inline docs/aws.md:74]: since swift-nio is a library this may be confusing. wither we pick another example, or at least lets change the command and call out explicitly that are we building the example  NIOHTTP1Server target which ships with swift-nio to make things clearer to a reader that is not familiar with those details

File: docs/aws.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+SwiftToolchainUrl="https://swift.org/builds/swift-5.4.1-release/amazonlinux2/swift-5.4.1-RELEASE/swift-5.4.1-RELEASE-amazonlinux2.tar.gz"
+sudo yum install ruby binutils gcc git glibc-static gzip libbsd libcurl libedit libicu libsqlite libstdc++-static libuuid libxml2 tar tzdata ruby -y
+cd $(mktemp -d)
+wget ${SwiftToolchainUrl} -O swift.tar.gz
+gunzip < swift.tar.gz | sudo tar -C / -xv --strip-components 1
+```
+
+Finally, check that Swift is correctly installed by running the Swift REPL: `swift`.
+
+![Invoke REPL](../images/aws/repl.png)
```
Comment (tomerd) [inline docs/aws.md:67]: would it be easier to do the build stage in Docker? instead of downloading and installing the toolchain, one could build in docker on their development machine (or an EC2 instance for development/build/CI purposes), then scp to the runtime EC2 instance.

File: docs/deployment.md (hunk 1 above)
Comment (gibbster) [reply docs/deployment.md]: Sure, I pushed an update that fixes this.

File: docs/deployment.md [hunk 4]
```diff
@@ -2,10 +2,11 @@
 ## Deployment to Servers or Public Cloud
 
 The following guides can help with the deployment to public cloud providers:
+* [AWS](aws.md)
 * [DigitalOcean](digital-ocean.md)
 * [Heroku](heroku.md)
 * [Docker/K8](packaging.md#docker)
```
Comment (tomerd) [inline docs/deployment.md:8]:
```suggestion
* [Kubernetes & Docker](packaging.md#docker)
```

File: docs/aws.md (hunk 2 above)
Comment (gibbster) [reply docs/aws.md:74]: I used swift-nio to parallel what was used in the ubuntu guide. I added a line of description (also matching the ubuntu guide) in the latest push

File: docs/aws.md (hunk 3 above)
Comment (gibbster) [reply docs/aws.md:67]: 🤷‍♂️  Both seem equally easy/hard to me. If you have a strong preference I can update it.

Comment (tomerd) [reply docs/aws.md:67]: @swift-server/sswg wdyt?

Comment (ktoso) [reply docs/aws.md:67]: dockering it up may be simpler for people to follow 👍 
A bit easier to point people at docker images than tell them what specific things to install every time.

**Review by ktoso — COMMENTED** [review]
Looks good, thanks a lot! :-)

I can see this being helpful if someone is first time ever going into aws and playing around with server side swift. Makes sense to include the guide I think.

File: docs/aws.md (hunk 3 above)
Comment (hassila) [reply docs/aws.md:67]: I’d just add another point of view - one might have a working ec2 setup but doesn’t use docker, which would make adding one more technology more complicated (my actual use case :-). That being said, either works of course, but just wanted to point out it adds another dimension of complexity.

Comment (ktoso) [reply docs/aws.md:67]: We could show both I guess? It’s not much more text to maintain tbh

Comment (0xTim) [reply docs/aws.md:67]: May view is that we should stick to compiling on the instance. Not everyone wants Docker and teaching people how to set up swift on a Linux box (even if it just points to the box) is not a bad thing.

IMO Docker should be a separate part of the guide and show the full lifecycle - compile container, push to ECR and deploy on to ECS/EKS etc

File: docs/aws.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+Click on "Launch Instance", either on the top of the screen, or if this is the first instance you have created in the region, in the main section of the screen.
+
+![Launch instance](../images/aws/launch-0.png)
+
+Choose an Amazon Machine Image (AMI). In this case the guide is assuming that we will be using Amazon Linux 2, so select that AMI type.
+
+![Choose AMI](../images/aws/launch-1.png)
+
+Choose an instance type. Larger instances types will have more memory and CPU, but will be more expensive. To experiment is it frugal to choose a smaller instance type. In this case I have a `t2.micro` instance type selected.
```
Comment (tachyonics) [inline docs/aws.md:23]: `frugal` seems like a very Amazon word. I know what you mean but it might be less obvious for someone else. Maybe something like-

```suggestion
Choose an instance type. Larger instances types will have more memory and CPU, but will be more expensive. A smaller instance type will be sufficient to experiment. In this case I have a `t2.micro` instance type selected.
```

File: docs/aws.md (hunk 3 above)
Comment (tachyonics) [reply docs/aws.md:67]: This particular guide seems focused on stepping through someone's first experimentation with Server Side Swift. Like Tim said, there is value in a very beginner level guide and local or cross-instance compilation feels like it would just be introducing a complication earlier than needed.

Comment (gibbster) [reply docs/aws.md:67]: It seems like there's a couple of different opinions :-) i.e. either reduce the moving parts by compiling on the instance, or use docker to simplify the build command. I'm happy to go in either direction, just let me know :-)

Comment (hassila) [reply docs/aws.md:67]: Friendly ping @tomerd ?

Comment (tomerd) [reply docs/aws.md:67]: how about we document both? this way folks that are familiar with Docker can go down that path, and others can install the toolchain on the host. hopefully we will have RPMs soon so the latter will be even simpler

Comment (gibbster) [reply docs/aws.md:67]: Yeah, I'm happy to do that. Give me a few days to wrap up another task on my plate and I'll get right on it.

Comment (gibbster) [reply docs/aws.md:71]: Just added a Docker option - I hope that's helpful

File: docs/aws.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
+# Deploying to AWS on Amazon Linux 2
```
Comment (tomerd) [inline docs/aws.md:1]: is this file here by mistake?

File: docs/allocations.md [hunk 7]
```diff
@@ -0,0 +1,343 @@
+# Allocations
```
Comment (tomerd) [inline docs/allocations.md:1]: is this file here by mistake?

File: docs/aws.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+Let's now download and build an test application. We will use the `--static-swift-stdlib` option so that it can be deployed to a different server without the Swift toolchain installed. These examples will deploy SwiftNIO's [example HTTP server](https://github.com/apple/swift-nio/tree/master/Sources/NIOHTTP1Server), but you can test with your own project.
+
+```
+git clone https://github.com/apple/swift-nio.git
+cd swift-nio
+swift build -v --static-swift-stdlib -c release
+```
+
+## Compile on Docker
```
Comment (tomerd) [inline docs/aws.md:81]:
```suggestion
## Compile with Docker
```

File: docs/aws.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+![Wait for instance launch and view details](../images/aws/ec2-list.png)
+
+Connect to instance. Using the keypair that you used or created in the launch step and the IP in the previous step, run ssh. Be sure to use the `-A` option with ssh so that in a future step we will be able to use the same key to connect to a second instance.
+
+![Connect to instance](../images/aws/ssh-0.png)
+
+We have two options to compile the binary: either directly on the instance or using Docker. We will go through both options here.
+
+## Compile on instance
+
```
Comment (tomerd) [inline docs/aws.md:58]: I think we can add a sentence here to highlight the two options to compile on the instance:
1. download and install the Swift toolchain 
2. use docker 

and then point / link to the relevant subsections

File: docs/linux-perf.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
+# Linux `perf`
```
Comment (tomerd) [inline docs/linux-perf.md:1]: is this file here by mistake?

**Comment by tomerd** [comment]
thanks @gibbster this is getting close, made some small suggestions.

also, looks like there are unrelated changes that ended up as part of this PR and need to be removed?

File: docs/performance.md [hunk 11]
```diff
@@ -103,6 +103,8 @@ for f in 0..<2_000 {
 
 The above program contains the `TerribleArray` data structure which has _O(n)_ appends and not the amortised _O(1)_ that users are used to from `Array`.
 
+We will assume, that you have Linux's `perf` installed and configured, documentation on how to install `perf` can be found in [this guide](linux-perf.md).
```
Comment (tomerd) [inline docs/performance.md:106]: these changes seem unrelated?

File: docs/setup-and-ide-alternatives.md [hunk 12]
```diff
@@ -7,6 +7,7 @@ The [supported platforms](https://swift.org/platform-support/) for running Swift
 A number of editors you may already be familiar with have support for writing Swift code. Here we provide a non-exhaustive list of such editors and relevant plugins/extensions, sorted alphabetically.
 
 * [Atom IDE support](https://atom.io/packages/ide-swift)
+    * [Atomic Blonde](https://atom.io/packages/atomic-blonde) a SourceKit based syntax highlighter for Atom.
```
Comment (tomerd) [inline docs/setup-and-ide-alternatives.md:10]: these changes seem unrelated?

Comment (gibbster) [reply docs/setup-and-ide-alternatives.md:10]: These seem to be related to this commit (not mine): https://github.com/swift-server/guides/commit/00340e7157bb1edf5339e1e8fc9091998a64ff03

File: docs/performance.md (hunk 11 above)
Comment (gibbster) [reply docs/performance.md]: Not sure how those two commits got in there. I reverted both

File: docs/aws.md (hunk 3 above)
Comment (ktoso) [reply docs/aws.md:71]: Thanks a lot, that's great 👍

File: docs/aws.md (hunk 9 above)
Comment (ktoso) [reply docs/aws.md:58]: +1 on that; it feels a bit weird now as we jump right into the downloading.

Maybe add this:


```suggestion
There are two alternative ways to compile code on the instance, either by:

- [downloading and using the toolchain directly on the instance](#compile-using-a-downloaded-toolchain),
- or by [using docker, and compiling inside a docker container](#compile-with-docker)

### Compile using a downloaded toolchain
```

Feel free to use this or use your own wording, then we're LGTM!

**Review by ktoso — APPROVED** [review]
Proposed a change but LGTM otherwise, thanks a lot it's a great addition to the guides!

**Review by tomerd — APPROVED** [review]
lgtm

**Comment by ktoso** [comment]
Thanks for the final change and thanks again the entire guide!
